### ISSUE TITLE
Variable not captured correctly

### ISSUE THREAD

**Issue by kubouch** [nushell/nushell#5994]
### Describe the bug

It seems like in some tricky cases the variables are not captured correctly.

### How to reproduce

Create a `.nu-env` file with the following contents:
```
def-env hook-shell [n: int, overlay: string] {
    let-env config = ($env.config | upsert hooks.env_change.NUSHELL_CURRENT_SHELL {
        [
            { let-env FOO = $n }
        ]
    })
}

hook-shell $env.NUSHELL_CURRENT_SHELL lol_shell
```
Then:
```
> source .nu-env
> # press enter
Error: nu::shell::variable_not_found (link)

  × Variable not found
   ╭─[.nu-env:3:1]
 3 │         [
 4 │             { let-env FOO = $n }
   ·                             ─┬
   ·                              ╰── variable not found
 5 │         ]
   ╰────
```

### Expected behavior

I would expect the `$n` variable to be captured from the argument of `hook-shell` and equal to the `$env.NUSHELL_CURRENT_SHELL`.

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.65.1                                          |
| branch             | hooks_string                                    |
| commit_hash        | eac5f62959c99a6ce3576d839f7f9b8a581d8486        |
| build_os           | linux-x86_64                                    |
| build_target       | x86_64-unknown-linux-gnu                        |
| rust_version       | rustc 1.62.0 (a8314ef7d 2022-06-27)             |
| rust_channel       | stable-x86_64-unknown-linux-gnu                 |
| cargo_version      | cargo 1.62.0 (a748cf5a3 2022-06-08)             |
| pkg_version        | 0.65.1                                          |
| build_time         | 2022-07-07 20:10:14 +03:00                      |
| build_rust_channel | release                                         |
| features           | database, dataframe, default, trash, which, zip |
| installed_plugins  | gstat                                           |

**Comment by kubouch** [comment]
Oh, oops, this is not meant to be closed. The error might be in the parser collecting the captures, I'm not sure.

**Comment by weirdan** [comment]
This looks fixed to me.

**Comment by kubouch** [comment]
I'm still seeing the same error.

**Comment by weirdan** [comment]
Are you still running it with `def-env`? I think that must be the problem.

With the updated syntax (`def --env`) this works:

<img width="1269" height="937" alt="Image" src="https://github.com/user-attachments/assets/f0a142f0-c249-4599-86af-e1b9a5063234" />

```nushell
{
  def --env hook-shell [n: string] {
    $env.config.hooks.env_change.NUSHELL_CURRENT_SHELL = [
      { $env.FOO = $n }
    ]
  }

  hook-shell $env.NUSHELL_CURRENT_SHELL
} 
| view source $in
| str trim --char '{'
| str trim --char '}'
| save --force .nu-env
```
```console
> $env.NUSHELL_CURRENT_SHELL = "abc"
> source .nu-env
> $env.FOO
```

I also tried it with `upsert`, as in the original report (to add more layers of closures) and it also worked fine.

**Comment by kubouch** [comment]
You're right, I forgot the `def-env`. With `def --env` it works. But it's weird that the parser complains about the `$n` and not `def-env`. Anyway, the issue indeed seems resolved.
